File: pipeline/naive/pipeline.nim
import std/[strformat]

const
  FnvOffset: uint64 = 0xcbf29ce484222325'u64
  FnvPrime: uint64  = 0x100000001b3'u64

proc fnv1a(s: string): uint64 =
  result = FnvOffset
  for c in s:
    result = (result xor uint64(ord(c))) * FnvPrime

var total: uint64 = 0
for i in 0 ..< 100000:
  let payload = &"task:item:{i}"
  total = total + fnv1a(payload)

echo &"Pipeline complete: processed=100000, checksum={total}"

File: pipeline/naive/pipeline.go
package main

import (
	"fmt"
	"strconv"
	"sync"
)

const (
	fnvOffset = 0xcbf29ce484222325
	fnvPrime  = 0x100000001b3
)

func fnv1a(str string) uint64 {
	h := uint64(fnvOffset)
	for i := 0; i < len(str); i++ {
		h = (h ^ uint64(str[i])) * fnvPrime
	}
	return h
}

type Task struct {
	id int
}

type Result struct {
	id       int
	checksum uint64
}

func main() {
	tasks := make(chan Task, 1000)
	results := make(chan Result, 1000)
	workers := 8
	var wg sync.WaitGroup

	for w := 0; w < workers; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for task := range tasks {
				payload := "task:item:" + strconv.Itoa(task.id)
				chk := fnv1a(payload)
				results <- Result{id: task.id, checksum: chk}
			}
		}()
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	go func() {
		for i := 0; i < 100000; i++ {
			tasks <- Task{id: i}
		}
		close(tasks)
	}()

	var total uint64
	count := 0
	for res := range results {
		total += res.checksum
		count++
	}

	fmt.Printf("Pipeline complete: processed=%d, checksum=%d\n", count, total)
}
